src/Twine/Guard.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+module Twine.Guard (
+    TerminationAction (..)
+  , TerminationHandler (..)
+  , guarded
+  , repeatedly
+  ) where
+
+import           Control.Monad.Catch
+import           Control.Monad.Trans.Either
+import           Control.Monad.Trans.Class
+
+import           P
+
+import           System.IO
+
+import           Twine.Snooze
+
+data TerminationAction =
+    Restart
+  | Die
+  deriving (Eq, Show)
+
+data TerminationHandler e =
+  TerminationHandler {
+      onExplosion :: SomeException -> IO TerminationAction
+    , onError :: e -> IO TerminationAction
+    , onGraceful :: IO TerminationAction
+    }
+
+-- | Run an action forever, using termination handler to notify of
```

Comment:
Ok, I was mistaken on the purpose of `Die` for the graceful handler then. I understand now.